### ISSUE TITLE
`prove_core` has unhelpful error message "Term is not a proposition"

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/saw-script#1566]
I have a polymorphic rewrite rule that I'd like to prove, which quantifies over a variable `n : Nat` that is used as a vector size parameter. It should be provable just by unfolding some definitions. But `prove_core` rejects it with an error message before even running the proof tactic, saying "Term is not a proposition".
```
prove_core
  do {
    simplify (cryptol_ss ());
    z3;
  }
  "\\ (n : Nat) -> \\ (x : Vec n Bool) -> Eq Integer (sbvToInt n x) (toSignedInteger (TCNum n) x)";
```
```
Stack trace:
"prove_core" (<stdin>:1:1-1:11):
termToProp: Term is not a proposition
\(n : Prelude.Nat) ->
  \(x : Prelude.Vec n Prelude.Bool) ->
    Prelude.Eq
      Prelude.Integer
      (Prelude.sbvToInt n x)
      (Cryptol.toSignedInteger (Cryptol.TCNum n) x)
(n : Prelude.Nat)
-> (x : Prelude.Vec n Prelude.Bool)
-> Prop
```

**Comment by robdockins** [comment]
I think you need to use Pi instead of lambda to state propositions.

**Comment by RyanGlScott** [comment]
Indeed, I ran into this issue in a separate context: https://github.com/aws/s2n-tls/commit/d5a044b8ce4eee8273397ee99d224faf8b562a39#diff-ae328ea2909d8141691e3a608d7acef67c9900d80e3c58a8ce08e341fb578d69R39

**Comment by robdockins** [comment]
You'll shortly run into a separate problem, though, which is that `Nat` isn't a first-order type.

**Comment by brianhuffman** [comment]
@robdockins Yes, of course, that's it. I was switching from `prove_print` (where you do need to use lambdas) to `prove_core` and forgot to switch from lambda to pi.

In that case, the bug is that the error message is unhelpful. I assumed it was complaining about quantification over `Nat` instead of the fact that it found a lambda instead of a pi.

**Comment by robdockins** [comment]
The following script just doesn't quite work.  We should probably make "trivial" a little less trivial:

```
prove_core
  do {
    simplify (cryptol_ss ());
    print_goal;
    trivial;
  }
  "(n : Nat) -> (x : Vec n Bool) -> Eq Integer (sbvToInt n x) (toSignedInteger (TCNum n) x)";
```

```
[18:56:07.043] Goal prove (goal number 0):
[18:56:07.043] (n : Nat)
-> (x : Vec n Bool)
-> let { x@1 = sbvToInt n x
    }
 in Eq Integer x@1 x@1
[18:56:07.043] Stack trace:
"prove_core" (/Users/rdockins/code/pate/test.saw:1:1-1:11):
"trivial" (/Users/rdockins/code/pate/test.saw:5:5-5:12):
trivial tactic: not a trivial goal
```

**Comment by brianhuffman** [comment]
Yes, it would be good to make `trivial` handle any proofs by reflexivity. I'll make a separate ticket for that. I think I can work around it for now.

**Comment by brianhuffman** [comment]
Here's my workaround, for the record:
```
fold_toSignedInteger <-
  prove_core
  do {
    simplify (addsimps [intEq_refl] ss0);
    trivial;
  }
  "(n : Nat) -> (x : Vec n Bool) -> EqTrue (intEq (sbvToInt n x) (toSignedInteger (TCNum n) x))";
```
where `intEq_refl` is a previously-proved rewrite rule for `intEq x x = True`.

**Comment by robdockins** [comment]
The current error is produced here:

https://github.com/GaloisInc/saw-script/blob/db538d30a650f9e89eb9410fcb58bc2f411f9a26/src/SAWScript/Proof.hs#L139-L148

It would probably not be hard to add a little extra step there to check if the leading symbol is a lambda and print an additional hint in that case.

**Comment by robdockins** [comment]
> Indeed, I ran into this issue in a separate context: [aws/s2n-tls@d5a044b#diff-ae328ea2909d8141691e3a608d7acef67c9900d80e3c58a8ce08e341fb578d69R39](https://github.com/aws/s2n-tls/commit/d5a044b8ce4eee8273397ee99d224faf8b562a39#diff-ae328ea2909d8141691e3a608d7acef67c9900d80e3c58a8ce08e341fb578d69R39)

Well, hummm.  I thought that https://github.com/GaloisInc/s2n/pull/7 had been pushed upstream.

**Comment by robdockins** [comment]
Fixed via #1568
